modules/zhttp/src/main/scala/com/caesars/tracing/TracedHttp.scala

```diff
@@ -0,0 +1,100 @@
+import sttp.model.HeaderNames
+import zhttp.http.*
+import zio.*
+
+object TracedHttp {
+  def apply[R](
+      ztracer: ZTracer,
+      dropRequestWhen: Request => Boolean = _ => false,
+      dropHeadersWhen: String => Boolean = HeaderNames.isSensitive
+  )(app: HttpApp[R, Throwable]): HttpApp[R, Throwable] = {
+    import TracedHttpUtils.*
+
+    Http.fromOptionFunction {
+      case request if !dropRequestWhen(request) => createUndroppedSpannedResponse(request, ztracer, app, dropHeadersWhen)
+      case request                              => app(request)
+    }
+  }
+
+  def layer[R](
+      dropRequestWhen: Request => Boolean = _ => false,
+      dropHeadersWhen: String => Boolean = HeaderNames.isSensitive
+  )(app: HttpApp[R, Throwable]): URIO[Has[ZTracer], HttpApp[R, Throwable]] =
+    ZIO.service[ZTracer].map(apply[R](_, dropRequestWhen, dropHeadersWhen)(app))
+}
+
+object TracedHttpUtils {
+  def putHeadersInSpan(res: Response, span: ZSpan, dropHeadersWhen: String => Boolean = HeaderNames.isSensitive): UIO[Unit] = {
+    val headers: List[(String, AttributeValue)] =
+      "status.code" -> LongValue(res.status.asJava.code().toLong) ::
+        "status.message" -> StringValue(res.status.asJava.reasonPhrase()) ::
+        res.headers.toList
+          .filterNot { case (k, _) => dropHeadersWhen(k) }
+          .map { case (k, v) => k -> StringValue(v) }
+
+    span.putAll(headers*)
+  }
+
+  def createSpannedResponse[R](
+      request: Request,
+      span: ZSpan,
+      app: HttpApp[R, Throwable],
+      dropHeadersWhen: String => Boolean = HeaderNames.isSensitive
+  ): ZIO[R, Option[Throwable], Response] = {
+    app(request).tapBoth(
+      {
+        case None    => span.setStatus(SpanStatus.NotFound)
+        case Some(e) => span.setStatus(SpanStatus.Internal(e.toString))
+      },
+      { res =>
+        span.setStatus(toSpanStatus(res.status)) *> putHeadersInSpan(res, span, dropHeadersWhen)
+      }
+    )
+  }
+
+  def createUndroppedSpannedResponse[R](
+      request: Request,
+      ztracer: ZTracer,
+      app: HttpApp[R, Throwable],
+      dropHeadersWhen: String => Boolean = HeaderNames.isSensitive
+  ): ZIO[R, Option[Throwable], Response] = {
+    val method = request.method
+    val url = request.url
+    val headers = request.headers
+    val name = s"$method ${url.path}"
+    val traceHeaders =
+      TraceHeaders.of(
+        headers.toList.filterNot { case (k, _) => dropHeadersWhen(k) }*
+      )
+
+    ztracer.fromHeaders(traceHeaders, SpanKind.Server, name)(createSpannedResponse(request, _, app, dropHeadersWhen))
+    // All errors on this effects should be converted into responses by the Http app composition
+    // e.g. app.catchAll(e => Response)
+    // otherwise, the fiber dies  and tracing is moot
+  }
+
+  implicit class StatusOps(val self: Status) extends AnyVal {
+    def isSuccess: Boolean =
+      self.asJava.code() >= 200 && self.asJava.code() < 400
```

Comment:
Looking at [http4s's handling](https://github.com/http4s/http4s/blob/main/core/shared/src/main/scala/org/http4s/Status.scala#L39-L44), please say why 3xx statuses mean success.

Reply:
Agreed, this is incomplete. Its only applicable in the context of tracing, where there is no `SpanStatus.Redirect`.